generator/src/main/java/com/scottlogic/deg/generator/generation/AutomatonUtils.java

```diff
@@ -11,54 +11,118 @@
+     * the longest possible variant.
+     *
+     * TODO:
+     * 1) turn this into a non-recursive algorithm
+     * 2) improve performance and/or memory management (don't build strings as part of the process)
+     *
+     * @param automaton The automaton that represents the regex
+     * @return The longest possible string for the given regex/automaton
+     */
     static String getLongestExample(Automaton automaton) {
+        List<Transition> transitions = automaton.getInitialState()
+            .getSortedTransitions(true)
+            .stream()
+            .sorted(Comparator.comparing(Transition::getMin))
+            .collect(Collectors.toList()); //ensure some form or determinism
+
+        TransitionLongestExampleCache cache = new TransitionLongestExampleCache();
+
+        String longest = "";
+        for (Transition transition: transitions) {
+            String transitionLongest = getLongestExample(transition, cache);
+            if (transitionLongest.length() > longest.length()) {
+                longest = transitionLongest;
+            }
+        }
+
+        return longest;
+    }
+
+    /**
+     * A cache of transitions to calculated strings, to save recalculating the string fragment for a transition multiple
+     * times.
+     */
+    private static class TransitionLongestExampleCache {
+        private final Map<Transition, String> examples = new HashMap<>();
+
+        String getLongestExample(Transition transition){
+            if (examples.containsKey(transition)){
+                return examples.get(transition);
+            }
+
+            String example = AutomatonUtils.getLongestExample(transition, this);
+            examples.put(transition, example);
+            return example;
+        }
+    }
 
-        Transition[] solution = {};
+    /**
+     * Recursive function for calculating the longest string that can be generated from the given regular automaton
+     *
+     * @param transition The transition to start at
+     * @param cache A cache of previously resolved transitions
+     * @return The longest string that can be generated from the given transition
+     */
+    private static String getLongestExample(Transition transition, TransitionLongestExampleCache cache){
+        if (transition == null){
+            throw new IllegalArgumentException("Transition cannot be null");
+        }
 
-        // The start node always has one transition to start
-        Transition startTransition = automaton.getInitialState().getSortedTransitions(true).get(0);
-        Transition[] startTransitionPath = { startTransition };
+        StringBuilder string = new StringBuilder();
 
-        Stack<Transition[]> transitionsToVisit = new Stack<>();
-        transitionsToVisit.add(startTransitionPath);
+        while (transition != null) { //need some exit condition
+            State dest = transition.getDest();
```

Comment:
what's a dest?

Reply:
The state a transition results in, not the best name - but keeping names here close to the automaton removes any ambiguity.

See this article which describes how the [Automaton represents and uses States and Transitions](https://github.com/ScottLogic/datahelix/blob/master/generator/docs/StringGeneration.md)